src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -1586,6 +1588,22 @@ private boolean bootstrap(final Collection<Token> tokens)
             SystemKeyspace.resetAvailableRanges();
         }
 
+        try
+        {
+            setMode(Mode.WAITING_TO_REQUEST_STREAMS, "Awaiting call to proceed with requesting streams during bootstrap", true);
+            boolean timeoutExceeded = !startRequestStreamsCondition.await(STREAMS_REQUEST_SAFETY_CHECK_GRACE_PERIOD_MINUTES, MINUTES);
+            if (timeoutExceeded)
+            {
+                logger.error("Start signal to request streams was not given within 30 minutes. Streams request safety check failed.");
+                recordBootstrapErrorAndThrow("streamsRequestSafetyCheckFailed");
+            }
+        }
```

Comment:
Could we add a unit test in `StorageServiceServerTest` to ensure we are already gossiping our tokens and JOINING application state at that point of the gate?